cfgmgr/coppmgr.cpp

```diff
@@ -0,0 +1,859 @@
+            {
+                FieldValueTuple fv(f.key(), f.value().get<std::string>());
+                fvs.push_back(fv);
+            }
+            if (table_name == CFG_COPP_TRAP_TABLE_NAME)
+            {
+                m_coppTrapInitCfg[table_key] = fvs;
+            }
+            else if (table_name == CFG_COPP_GROUP_TABLE_NAME)
+            {
+                m_coppGroupInitCfg[table_key] = fvs;
+            }
+        }
+    }
+}
+
+/* Check if the trap group has traps that can be installed only when
+ * feature is enabled
+ */
+bool CoppMgr::checkIfTrapGroupFeaturePending(string trap_group_name)
+{
+    std::vector<FieldValueTuple> fvs;
+
+    if (m_coppGroupRestrictedMap.find(trap_group_name) == m_coppGroupRestrictedMap.end())
+    {
+        return false;
+    } 
+
+    for (auto it: m_coppTrapIdTrapGroupMap)
+    {
+        if (it.second == trap_group_name)
+        {
+            /* At least one trap should be enabled to install the 
+             * trap group with restricted attributes
+             */ 
+            if (!isTrapDisabled(it.first))
+            {
+                return false;
+            }
+        }
+    }
+    return true;
+}
+
+void CoppMgr::setFeatureTrapIdsStatus(string feature, bool enable)
+{
+    for (auto it: trap_feature_map)
+    {
+        if (it.second ==  feature)
+        {
+            if (enable)
+            {
+                if (m_coppTrapDisabledMap.find(it.first) != m_coppTrapDisabledMap.end())
+                {
+                    if (m_coppTrapIdTrapGroupMap.find(it.first) != m_coppTrapIdTrapGroupMap.end())
+                    {
+                        string trap_group = m_coppTrapIdTrapGroupMap[it.first];
+
+                        /* Enable only when the restricted group is already pending */
+                        if ((m_coppGroupRestrictedMap.find(trap_group) 
+                             != m_coppGroupRestrictedMap.end()) &&
+                             (checkIfTrapGroupFeaturePending(trap_group)))
+                        {
+                            vector<FieldValueTuple> fvs;
+                            vector<FieldValueTuple> modified_fvs;
+                            string trap_ids;
+
+                            for (auto i : m_coppGroupRestrictedMap[trap_group])
+                            {
+                                FieldValueTuple fv(i.first, i.second);
+                                fvs.push_back(fv);
+                            }
+                            getTrapGroupTrapIds(trap_group, trap_ids);
+                            if (!trap_ids.empty())
+                            {
+                                FieldValueTuple fv(COPP_TRAP_ID_LIST_FIELD, trap_ids);
+                                fvs.push_back(fv);
+                            }
+                            coppGroupGetModifiedFvs (trap_group, fvs, modified_fvs, false);
+                            if (!modified_fvs.empty())
```

Comment:
Not sure I understand this. Why do we check for modified_fvs? If a feature is enabled, just add that to the trap_id list right?

Reply:
Same as explained above. The modified fvs would program either the full table if not already programmed or trap ids if only trap ids change